sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala

```diff
@@ -0,0 +1,227 @@
+    }
+  }
+
+  private def buildReaderBase[T](
+      file: PartitionedFile,
+      buildReaderFunc: (
+        ParquetInputSplit, InternalRow, TaskAttemptContextImpl, Option[FilterPredicate],
+          Option[TimeZone]) => RecordReader[Void, T]): RecordReader[Void, T] = {
+    val conf = broadcastedConf.value.value
+
+    val filePath = new Path(new URI(file.filePath))
+    val split =
+      new org.apache.parquet.hadoop.ParquetInputSplit(
+        filePath,
+        file.start,
+        file.start + file.length,
+        file.length,
+        Array.empty,
+        null)
+
+    lazy val footerFileMetaData =
+      ParquetFileReader.readFooter(conf, filePath, SKIP_ROW_GROUPS).getFileMetaData
+    // Try to push down filters when filter push-down is enabled.
+    val pushed = if (enableParquetFilterPushDown) {
+      val parquetSchema = footerFileMetaData.getSchema
+      val parquetFilters = new ParquetFilters(parquetSchema, pushDownDate, pushDownTimestamp,
+        pushDownDecimal, pushDownStringStartWith, pushDownInFilterThreshold, isCaseSensitive)
+      filters
+        // Collects all converted Parquet filter predicates. Notice that not all predicates can be
+        // converted (`ParquetFilters.createFilter` returns an `Option`). That's why a `flatMap`
+        // is used here.
+        .flatMap(parquetFilters.createFilter)
+        .reduceOption(FilterApi.and)
+    } else {
+      None
+    }
+    // PARQUET_INT96_TIMESTAMP_CONVERSION says to apply timezone conversions to int96 timestamps'
+    // *only* if the file was created by something other than "parquet-mr", so check the actual
+    // writer here for this file.  We have to do this per-file, as each file in the table may
+    // have different writers.
+    // Define isCreatedByParquetMr as function to avoid unnecessary parquet footer reads.
+    def isCreatedByParquetMr: Boolean =
+      footerFileMetaData.getCreatedBy().startsWith("parquet-mr")
+
+    val convertTz =
+      if (timestampConversion && !isCreatedByParquetMr) {
+        Some(DateTimeUtils.getTimeZone(conf.get(SQLConf.SESSION_LOCAL_TIMEZONE.key)))
+      } else {
+        None
+      }
+
+    val attemptId = new TaskAttemptID(new TaskID(new JobID(), TaskType.MAP, 0), 0)
+    val hadoopAttemptContext = new TaskAttemptContextImpl(conf, attemptId)
+
+    // Try to push down filters when filter push-down is enabled.
+    // Notice: This push-down is RowGroups level, not individual records.
+    if (pushed.isDefined) {
+      ParquetInputFormat.setFilterPredicate(hadoopAttemptContext.getConfiguration, pushed.get)
+    }
+    val reader =
+      buildReaderFunc(split, file.partitionValues, hadoopAttemptContext, pushed, convertTz)
+    reader.initialize(split, hadoopAttemptContext)
+    reader
+  }
+
+  private def createRowBaseReader(file: PartitionedFile): ParquetRecordReader[UnsafeRow] = {
+    buildReaderBase(file, createRowBaseReader0).asInstanceOf[ParquetRecordReader[UnsafeRow]]
```

Comment:
`buildReaderBase` is parameterized, but the result is still casted. Why not parameterize so that ti retursn `ParquetRecordReader[UnsafeRow]` to avoid the cast? I think these casts should be removed.